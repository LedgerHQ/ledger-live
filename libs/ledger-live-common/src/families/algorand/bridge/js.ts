import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import type { AlgorandTransaction } from "../types";
import { sync, scanAccounts } from "../js-synchronization";
import {
  createTransaction,
  prepareTransaction,
} from "../js-prepareTransaction";
import { estimateMaxSpendable } from "../js-estimateMaxSpendable";
import { getTransactionStatus } from "../js-getTransactionStatus";
import { signOperation } from "../js-signOperation";
import { broadcast } from "../js-broadcast";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const updateTransaction = (t, patch) => {
  return { ...t, ...patch };
};

const preload = async () => Promise.resolve({});

const hydrate = () => {};

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<AlgorandTransaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast,
  estimateMaxSpendable,
};

export default {
  currencyBridge,
  accountBridge,
};
