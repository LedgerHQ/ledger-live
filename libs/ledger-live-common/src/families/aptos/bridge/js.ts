import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import { getSerializedAddressParameters } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../synchronisation";
import getTransactionStatus from "../getTransactionStatus";
import prepareTransaction from "../prepareTransaction";
import createTransaction from "../createTransaction";
import estimateMaxSpendable from "../estimateMaxSpendable";
import signOperation from "../signOperation";
import broadcast from "../broadcast";

const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};

const updateTransaction = (t: Transaction, patch: Partial<Transaction>): Transaction => ({
  ...t,
  ...patch,
});

const receive = makeAccountBridgeReceive();

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  getSerializedAddressParameters,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
