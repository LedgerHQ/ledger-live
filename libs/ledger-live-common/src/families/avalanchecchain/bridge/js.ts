import type { AccountBridge, CurrencyBridge } from "../../../types";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import { createTransaction, prepareTransaction, updateTransaction } from '../js-transaction';
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from '../js-signOperation';
import broadcast from "../js-broadcast";
import estimateMaxSpendable from '../js-estimateMaxSpendable';

const preload = () => Promise.resolve({});
const hydrate = (): void => { };

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
    preload,
    hydrate,
    scanAccounts,
};
const accountBridge: AccountBridge<Transaction> = {
    estimateMaxSpendable,
    createTransaction,
    updateTransaction,
    getTransactionStatus,
    prepareTransaction,
    sync,
    receive,
    signOperation,
    broadcast,
};
export default {
    currencyBridge,
    accountBridge,
};
