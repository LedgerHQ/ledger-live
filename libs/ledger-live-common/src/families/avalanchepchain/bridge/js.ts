import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import {
  createTransaction,
  prepareTransaction,
  updateTransaction,
} from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { preload, hydrate } from "../js-preload-data";
import estimateMaxSpendable from "../js-estimateMaxSpendable";

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
export default {
  currencyBridge,
  accountBridge,
};
