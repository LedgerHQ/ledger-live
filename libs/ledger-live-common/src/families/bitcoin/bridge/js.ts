import type { Transaction } from "../types";
import { sync, scanAccounts } from "../js-synchronisation";
import createTransaction from "../js-createTransaction";
import prepareTransaction from "../js-prepareTransaction";
import getTransactionStatus from "../js-getTransactionStatus";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { calculateFees } from "./../cache";
import { perCoinLogic } from "../logic";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import * as explorerConfigAPI from "../../../api/explorerConfig";
import { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive({
  injectGetAddressParams: (account) => {
    const perCoin = perCoinLogic[account.currency.id];

    if (perCoin && perCoin.injectGetAddressParams) {
      return perCoin.injectGetAddressParams(account);
    }
  },
});

const updateTransaction = (t, patch) => {
  const updatedT = { ...t, ...patch };

  // We accept case-insensitive addresses as input from user,
  // but segwit addresses need to be converted to lowercase to be valid
  if (updatedT.recipient.toLowerCase().indexOf("bc1") === 0) {
    updatedT.recipient = updatedT.recipient.toLowerCase();
  }

  return updatedT;
};

const preload = async () => {
  const explorerConfig = await explorerConfigAPI.preload();
  return {
    explorerConfig,
  };
};

const hydrate = (maybeConfig: any) => {
  if (
    typeof maybeConfig === "object" &&
    maybeConfig &&
    maybeConfig.explorerConfig
  ) {
    explorerConfigAPI.hydrate(maybeConfig.explorerConfig);
  }
};

const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload,
  hydrate,
};
const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  prepareTransaction,
  updateTransaction,
  getTransactionStatus,
  receive,
  sync,
  signOperation,
  broadcast: async ({ account, signedOperation }) => {
    calculateFees.reset();
    return broadcast({
      account,
      signedOperation,
    });
  },
};
export default {
  currencyBridge,
  accountBridge,
};
