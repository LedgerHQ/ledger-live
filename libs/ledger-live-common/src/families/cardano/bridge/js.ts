import type { Transaction } from "../types";
import { scanAccounts, sync } from "../js-synchronisation";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import {
  createTransaction,
  prepareTransaction,
  updateTransaction,
} from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: async () => ({}),
  hydrate: () => {},
};

export default { currencyBridge, accountBridge };
