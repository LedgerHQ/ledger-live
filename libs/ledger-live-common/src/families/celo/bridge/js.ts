import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import { getPreloadStrategy, preload, hydrate } from "../preload";
import signOperation from "../js-signOperation";
import getTransactionStatus from "../js-getTransactionStatus";
import broadcast from "../js-broadcast";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import prepareTransaction from "../js-prepareTransaction";
import createTransaction from "../js-createTransaction";

const updateTransaction = (t, patch) => ({ ...t, ...patch });

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  getPreloadStrategy,
  preload,
  hydrate,
  scanAccounts,
};
const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
export default {
  currencyBridge,
  accountBridge,
};
