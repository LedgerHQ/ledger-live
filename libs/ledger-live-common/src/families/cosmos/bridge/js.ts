import createTransaction from "../js-createTransaction";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import getTransactionStatus from "../js-getTransactionStatus";
import prepareTransaction from "../js-prepareTransaction";
import signOperation from "../js-signOperation";
import { sync, scanAccounts } from "../js-synchronisation";
import updateTransaction from "../js-updateTransaction";
import type { CosmosValidatorItem, Transaction } from "../types";
import cosmosValidatorsManager from "../validators";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { defaultCosmosAPI } from "../api/Cosmos";
import {
  asSafeCosmosPreloadData,
  setCosmosPreloadData,
} from "../preloadedData";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const getPreloadStrategy = (_currency) => ({
  preloadMaxAge: 30 * 1000,
});

const currencyBridge: CurrencyBridge = {
  getPreloadStrategy,
  preload: async () => {
    const validators = await cosmosValidatorsManager.getValidators();
    setCosmosPreloadData({
      validators,
    });
    return Promise.resolve({
      validators,
    });
  },
  hydrate: (data: { validators?: CosmosValidatorItem[] }) => {
    if (!data || typeof data !== "object") return;
    const { validators } = data;
    if (
      !validators ||
      typeof validators !== "object" ||
      !Array.isArray(validators)
    )
      return;
    cosmosValidatorsManager.hydrateValidators(validators);
    setCosmosPreloadData(asSafeCosmosPreloadData(data));
  },
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  estimateMaxSpendable,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast: defaultCosmosAPI.broadcast,
};

export default {
  currencyBridge,
  accountBridge,
};
