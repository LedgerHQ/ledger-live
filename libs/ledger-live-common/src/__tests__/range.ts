import { BigNumber } from "bignumber.js";
import {
  fromRangeRaw,
  toRangeRaw,
  inferDynamicRange,
  projectRangeIndex,
  reverseRangeIndex,
} from "../range";

function asString(bn) {
  if (BigNumber.isBigNumber(bn)) return bn.toString();

  if (typeof bn === "object" && bn) {
    const obj = {};

    for (const k in bn) {
      obj[k] = asString(bn[k]);
    }

    return obj;
  }

  return bn;
}

test("inferDynamicRange", () => {
  expect(asString(inferDynamicRange(new BigNumber(0.4)))).toMatchSnapshot();
  expect(asString(inferDynamicRange(new BigNumber(1)))).toMatchSnapshot();
  expect(asString(inferDynamicRange(new BigNumber(1.01)))).toMatchSnapshot();
  expect(asString(inferDynamicRange(new BigNumber(10)))).toMatchSnapshot();
  expect(asString(inferDynamicRange(new BigNumber(100)))).toMatchSnapshot();
  expect(asString(inferDynamicRange(new BigNumber(99)))).toMatchSnapshot();
});

test("projectRangeIndex", () => {
  expect(
    asString(projectRangeIndex(inferDynamicRange(new BigNumber(0.4)), 2))
  ).toMatchSnapshot();
});

test("reverseRangeIndex", () => {
  const range = inferDynamicRange(new BigNumber(0.4));

  for (let i = 0; i < range.steps; i++) {
    const n = projectRangeIndex(range, i);
    expect(reverseRangeIndex(range, n)).toBe(i);
  }
});

describe("RangeRaw", () => {
  [0.4, 1, 10, 111, 666, 99999].forEach((v) => {
    test("fromRangeRaw(toRangeRaw(x)) is identity for inferred " + v, () => {
      const range = inferDynamicRange(new BigNumber(v));
      expect(fromRangeRaw(toRangeRaw(range))).toEqual(range);
    });
  });
});
