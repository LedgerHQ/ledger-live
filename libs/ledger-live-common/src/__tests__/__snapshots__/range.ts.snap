// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inferDynamicRange 1`] = `
{
  "initial": "0.4",
  "max": "0.8",
  "min": "0.12",
  "step": "0.005",
  "steps": 137,
}
`;

exports[`inferDynamicRange 2`] = `
{
  "initial": "1",
  "max": "2",
  "min": "0.3",
  "step": "0.01",
  "steps": 171,
}
`;

exports[`inferDynamicRange 3`] = `
{
  "initial": "1.01",
  "max": "2.02",
  "min": "0.31",
  "step": "0.01",
  "steps": 172,
}
`;

exports[`inferDynamicRange 4`] = `
{
  "initial": "10",
  "max": "20",
  "min": "3",
  "step": "0.1",
  "steps": 171,
}
`;

exports[`inferDynamicRange 5`] = `
{
  "initial": "100",
  "max": "200",
  "min": "30",
  "step": "1",
  "steps": 171,
}
`;

exports[`inferDynamicRange 6`] = `
{
  "initial": "99",
  "max": "198",
  "min": "30",
  "step": "1",
  "steps": 169,
}
`;

exports[`projectRangeIndex 1`] = `"0.13"`;
